ikarus/finiteelements/mechanics/materials/vanishingstrain.hh

```diff
@@ -0,0 +1,180 @@
+// SPDX-FileCopyrightText: 2021-2024 The Ikarus Developers mueller@ibb.uni-stuttgart.de
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+/**
+ * \file vanishingstrain.hh
+ * \brief Defines the VanishingStrain material model and related functions.
+ * \ingroup  materials
+ */
+
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+#pragma once
+
+#include "vanishinghelpers.hh"
+
+#include <ikarus/finiteelements/mechanics/materials/interface.hh>
+#include <ikarus/finiteelements/mechanics/materials/strainconversions.hh>
+#include <ikarus/solver/nonlinearsolver/newtonraphson.hh>
+#include <ikarus/utils/nonlinearoperator.hh>
+
+namespace Ikarus {
+
+/**
+ * \brief VanishingStrain material model that enforces strain components to be zero.
+ * \ingroup materials
+ * \tparam strainIndexPair An array of MatrixIndexPair representing fixed stress components.
+ * \tparam MI The underlying material model.
+ */
+template <auto strainIndexPair, typename MI>
+struct VanishingStrain : public Material<VanishingStrain<strainIndexPair, MI>>
+{
+  using Underlying         = MI;                              ///< The underlying material type.
+  using ScalarType         = typename Underlying::ScalarType; ///< Scalar type.
+  using MaterialParameters = typename Underlying::MaterialParameters;
+
+  static constexpr auto fixedPairs        = strainIndexPair;                     ///< Array of fixed stress components.
+  static constexpr auto freeVoigtIndices  = createfreeVoigtIndices(fixedPairs);  ///< Free Voigt indices.
+  static constexpr auto fixedVoigtIndices = createFixedVoigtIndices(fixedPairs); ///< Fixed Voigt indices.
+  static constexpr auto freeStrains       = freeVoigtIndices.size();             ///< Number of free strains.
+
+  static constexpr auto strainTag          = Underlying::strainTag;          ///< Strain tag.
+  static constexpr auto stressTag          = Underlying::stressTag;          ///< Stress tag.
+  static constexpr auto tangentModuliTag   = Underlying::tangentModuliTag;   ///< Tangent moduli tag.
+  static constexpr bool energyAcceptsVoigt = Underlying::energyAcceptsVoigt; ///< Energy accepts Voigt notation.
+  static constexpr bool stressToVoigt      = true;                           ///< Stress to Voigt notation.
+  static constexpr bool stressAcceptsVoigt = true;                           ///< Stress accepts Voigt notation.
+  static constexpr bool moduliToVoigt      = true;                           ///< Moduli to Voigt notation.
+  static constexpr bool moduliAcceptsVoigt = true;                           ///< Moduli accepts Voigt notation.
+  static constexpr double derivativeFactor = 1;                              ///< Derivative factor.
```

Comment:
This might be wrong. Also in the vanishing stress case. This factor tells if the derivatives from energy to stress and to material tangent are related by a factor of one or two.

Reply:
Should we then use just `Underlying::derivativeFactor` ?

Reply:
Yes